ssl/ssl_lib.c

```diff
@@ -3847,6 +3848,55 @@ static int ssl_session_cmp(const SSL_SESSION *a, const SSL_SESSION *b)
     return memcmp(a->session_id, b->session_id, a->session_id_length);
 }
 
+#ifndef OPENSSL_NO_SSLKEYLOG
+/**
+ * @brief Static initialization for a one-time action to initialize the SSL key log.
+ */
+static CRYPTO_ONCE ssl_keylog_once = CRYPTO_ONCE_STATIC_INIT;
+
+/**
+ * @brief Pointer to a read-write lock used to protect access to the key log.
+ */
+static CRYPTO_RWLOCK *keylog_lock = NULL;
+
+/**
+ * @brief Pointer to a BIO structure used for writing the key log information.
+ */
+static BIO *keylog_bio = NULL;
+
+/**
+ * @brief Ref counter tracking the number of keylog_bio users.
+ */
+static unsigned int keylog_count = 0;
+
+/**
+ * @brief Initializes the SSLKEYLOGFILE lock.
+ *
+ * @return 1 on success, 0 on failure.
+ */
+DEFINE_RUN_ONCE_STATIC(ssl_keylog_init)
+{
+    keylog_lock = CRYPTO_THREAD_lock_new();
```

Comment:
Why not just initialise the BIO here and do away with all the ref counting etc? Or is it because we want to explicitly close the BIO if no one is using it for a while (although if that is the case I'm not really sure that we have achieved it with this code - since the SSL_CTX is generally expected to be long-living).

Reply:
yes, exactly.  Its possible we will have multiple SSL_CTX objects using the global keylog file, and so we need to track how many users there are, and only free it when that counter hits zero.  We could of course initalize it in the run_once routine, but that would:
1) Not eliminate the need for a reference counter
2) Make it hard for situations in which a user opens one more more SSL_CTX's, closes them all (which would close the bio), and then open another SSL_CTX (the last step would not then have a keylog_bio if we placed its creation in the run_once routine